app/modules/collaborations/models.py

```diff
@@ -17,13 +51,161 @@ class Collaboration(db.Model, HoustonModel):
+        for association in self.collaboration_user_associations:
+            if association.read_approval_state == CollaborationUserState.DECLINED:
+                # only one user need decline
+                read_state = CollaborationUserState.DECLINED
+                break
+            elif association.read_approval_state == CollaborationUserState.PENDING:
+                # only one user needs pending, but we should still keep looking for declined
+                read_state = CollaborationUserState.PENDING
+            elif (
+                association.read_approval_state == CollaborationUserState.APPROVED
+                and read_state is None
+            ):
+                # every association must approve to return approved, everything else overrides it
+                read_state = CollaborationUserState.APPROVED
+        return read_state
+
+    def get_edit_state(self):
+        edit_state = None
+        for association in self.collaboration_user_associations:
+            # you should never have a 'pending' and 'not_initiated' on the same collab, so one of either is enough
+            if association.edit_approval_state == CollaborationUserState.DECLINED:
+                edit_state = CollaborationUserState.DECLINED
+                break
+            elif association.edit_approval_state == CollaborationUserState.NOT_INITIATED:
+                edit_state = CollaborationUserState.NOT_INITIATED
+                break
+            elif association.edit_approval_state == CollaborationUserState.PENDING:
+                edit_state = CollaborationUserState.PENDING
+            elif (
+                association.read_approval_state == CollaborationUserState.APPROVED
+                and edit_state is None
+            ):
+                edit_state = CollaborationUserState.APPROVED
+        return edit_state
+
+    def set_read_approval_state_for_user(self, user_guid, state):
+        from app.modules.users.models import User
+
+        if (
+            user_guid is not None
+            and User.ensure_edm_obj(user_guid) is not None
+            and state in CollaborationUserState.ALLOWED_STATES
+        ):
+            for association in self.collaboration_user_associations:
+                if association.user_guid == user_guid:
+                    association.read_approval_state = state
+
+    def set_edit_approval_state_for_user(self, user_guid, state):
+        from app.modules.users.models import User
+
+        if user_guid is not None and User.ensure_edm_obj(user_guid) is not None:
+            # if one association is edit level NOT_INITIATED, they all are
+            if (
+                self.get_edit_state() == CollaborationUserState.NOT_INITIATED
+                and state is not CollaborationUserState.NOT_INITIATED
+            ):
+                self.initate_edit_with_user(user_guid)
+            for association in self.collaboration_user_associations:
```

Comment:
duplication twice is ok, at three times I get twitchy. 
at this point it would probably help to write a _get_association_for_user(user_guid) that iterated through self.collaboration_user_associations and returned either the assocation of None then ditch the for loop in these three functions